### DISCUSSION TITLE
"You provided undefined were a stream was expected" is extremely hard to debug

### DISCUSSION THREAD

**Discussion by ShacharHarshuv** [ReactiveX/rxjs#6788] [Ideas / Feature request]
Consider the following code: 
```js
const myObservable$ = combineLatest([someObservable$]);
```
This code will not fail even if `someObservable$` is undefined, but it WILL fail while trying to subscribe it:
```js
const myObservable$ = combineLatest([someObservable$]);
myObservable$.subscribe(); // this will fail
```

That makes it extremely difficult to debug and find where is the place were undefined was provided were a stream was expected. Especially if there's a long change with combineLatests spread across different file, and the subscribe call is in entirely different place. 

I feel like **a good solution for this is if combineLatest had thrown an error immediately when called with illegal values**, so the call-stack will show the exact place the error is in, and will make it easier to debug.

I suspect it's not a problem ONLY with combineLatest, and of course the solution should address all situation were this error could occur.

What do you think?

**Comment by voliva** [comment]
Yep - I think that would improve it.

However, the place where I've had this error happening most of the time is within a switchMap:

```ts
const myObservable$ = of(1).pipe(
  switchMap(v => getObservableForValue(v))
)
```

If `getObservableForValue` returns `undefined`, then this will blow up with that exception, and there's no way of detecting it (unless using when typescript)

Essentially this happens with all high order observables - `combineLatest` can also be written with `of({ list of observables }) + combineLatestAll()` - On that case, `of` can't throw if one of the values is undefined, but then combineLatestAll will.

**Reply by ShacharHarshuv** [reply]
The goal here is to throw the error before subscribing, and I don't think this is possible in those cases:

1. For switchMap - there is no way to know what will be to returned value of the switchMap callback for every input.
2. For combineLatestAll - you cannot pre-detemined what will be the emissions of the source observable without subscribing to it.

However this will work for combineLatest because you have all the values in the moment of calling the function. (I.e. before subscribing)

**Comment by ShacharHarshuv** [comment]
Any other input...? I'd like to implement this but I first want to know of there's an objection / an issue I can't see.
